1112-i386-txt-Add-Intel-TXT-ACM-module-support.patch

```diff
@@ -0,0 +1,596 @@
++       !grub_txt_acmod_match_platform (bios_sinit) )
++    return NULL;
++
++  return bios_sinit;
++}
++
++grub_uint32_t
++grub_txt_supported_os_sinit_data_ver (struct grub_txt_acm_header* hdr)
++{
++  static struct grub_txt_acm_info_table *info_table;
++
++  /* Assumes that it passed is_sinit_acmod() */
++  info_table = get_acmod_info_table (hdr);
++
++  if ( info_table == NULL )
++    return 0;
++
++  return info_table->os_sinit_data_ver;
++}
++
++grub_uint32_t
++grub_txt_get_sinit_capabilities (struct grub_txt_acm_header* hdr)
++{
++  static struct grub_txt_acm_info_table *info_table;
++
++  /* Assumes that it passed is_sinit_acmod() */
++  info_table = get_acmod_info_table (hdr);
++
++  if ( info_table == NULL || info_table->version < 3 )
++    return 0;
++
++  return info_table->capabilities;
++}
++
++int
++grub_txt_is_sinit_acmod (const void *acmod_base, grub_uint32_t acmod_size)
++{
++  grub_uint8_t type;
++
++  if ( !is_acmod (acmod_base, acmod_size, &type) )
++    return 0;
++
++  if ( type != GRUB_TXT_ACM_CHIPSET_TYPE_SINIT )
++    return 0;
++
++  return 1;
++}
++
++/* Format of VER.FSBIF and VER.QPIIF registers. */
++typedef union {
++  grub_uint64_t _raw;
++  struct {
++     grub_uint64_t reserved   : 31;
++     grub_uint64_t prod_fused : 1;
++  };
++} grub_txt_ver_fsbif_qpiif_t;
++
++int
++grub_txt_acmod_match_platform (struct grub_txt_acm_header *hdr)
++{
++  union grub_txt_didvid didvid;
++  grub_uint32_t fms, ign, i;
++  grub_uint64_t platform_id;
++  grub_txt_ver_fsbif_qpiif_t ver;
++  struct grub_txt_acm_chipset_id_list *chipset_id_list;
++  struct grub_txt_acm_chipset_id *chipset_id;
++  struct grub_txt_acm_processor_id_list *proc_id_list;
++  struct grub_txt_acm_processor_id *proc_id;
++  struct grub_txt_acm_info_table *info_table;
++
++  /* This fn assumes that the ACM has already passed the is_acmod() checks */
++
++  /* Get chipset fusing, device, and vendor id info */
++  didvid.value = grub_txt_reg_pub_readq (GRUB_TXT_DIDVID);
++
++  ver._raw = grub_txt_reg_pub_readq (GRUB_TXT_VER_FSBIF);
++  if ( (ver._raw & 0xffffffff) == 0xffffffff ||
++       (ver._raw & 0xffffffff) == 0x00 ) /* Need to use VER.QPIIF */
++    ver._raw = grub_txt_reg_pub_readq (GRUB_TXT_VER_QPIIF);
++
++  grub_dprintf ("slaunch", "chipset production fused: %x, "
++		"chipset vendor: 0x%x, device: 0x%x, revision: 0x%x\n",
++		ver.prod_fused, didvid.vid, didvid.did, didvid.rid);
++
++  grub_cpuid (1, fms, ign, ign, ign);
++  platform_id = grub_rdmsr (GRUB_MSR_X86_PLATFORM_ID);
++
++  grub_dprintf ("slaunch", "processor family/model/stepping: 0x%x, "
++		"platform id: 0x%" PRIxGRUB_UINT64_T "\n", fms, platform_id);
++
++  /*
++   * Check if chipset fusing is same. Note the DEBUG.FUSE bit in the version
++   * is 0 when debug fused so the logic below checking a mismatch is valid.
++   */
++  if ( (ver._raw & GRUB_TXT_VERSION_DEBUG_FUSED) &&
++       (hdr->flags & GRUB_TXT_ACM_FLAG_DEBUG_SIGNED) )
```

Comment:
This errors out on a production chipset with a debug ACM, but not on a debug chipset with a production ACM. I don't think that's intended.